recipe-client-addon/lib/Sampling.jsm

```diff
@@ -130,4 +130,35 @@ this.Sampling = {
 
     return Sampling.isHashInBucket(inputHash, wrappedStart, end, total);
   }),
+
+  /**
+   * Sample over a list of ratios such that, over the input space, each ratio
+   * has a number of matches in correct proportion to the other ratios.
```

Comment:
Also, I wonder if this would be simpler as its own PR - it doesn't seem strictly related to the preferences API?